rust/private/rust.bzl

```diff
@@ -118,6 +118,48 @@ def crate_root_src(attr, srcs, crate_type):
         fail("No {} source file found.".format(" or ".join(file_names)), "srcs")
     return crate_root
 
+def _invalid_chars_in_crate_name(name):
+    """Returns any invalid chars in the given crate name.
+
+    Args:
+        name (str): Name to test.
+
+    Returns:
+        list: List of invalid characters in the crate name.
+    """
+
+    return dict([(c, ()) for c in name.elems() if not (c.isalnum() or c == "_")]).keys()
+
+def _crate_name(label, attr):
+    """Returns the crate name to use for the current target.
+
+    Args:
+        label (Label): The label of the current target.
+        attr (struct): The attributes of the current target.
+
+    Returns:
+        str: The crate name to use for this target.
+    """
+    if attr.crate_name:
+        invalid_chars = _invalid_chars_in_crate_name(attr.crate_name)
+        if invalid_chars:
+            fail("Crate name '{}' contains invalid character(s): {}".format(
+                attr.crate_name,
+                " ".join(invalid_chars),
+            ))
+        return attr.crate_name
+
+    crate_name = name_to_crate_name(label.name)
+    invalid_chars = _invalid_chars_in_crate_name(crate_name)
+    if invalid_chars:
+        fail(
+            "Crate name '{}'".format(crate_name),
```

Comment:
I think you want to be `+`ing these, rather than `,`ing them together?

Reply:
Thanks -- done!